### PR TITLE
Core: don't allow region, location, or entrance with duplicate names

### PR DESCRIPTION
**Pull request by alwaysintreble**
## What is this fixing or adding?
Title.

## How was this tested?
tested the worlds that are being changed in this PR.

### CODE REVIEW

**Comment by alwaysintreble** [comment]
games currently broken as of the initial writing of this assert:
* hylics 2
* lingo
* mega man battle network 3
* overcooked 2
* pokemon red and blue
* risk of rain 2
* super mario 64
* timespinner
* VVVVVV

**Comment by ThePhar** [comment]
@Alchav explicit ping for you since you didn't accept the collaborator invite so I can't request review from you 😿.

**Review by toasterparty — APPROVED** [review]

**Review by kindasneaki — APPROVED** [review]
This makes sense why I couldn't add a rule on a different entrance.

File: worlds/sm64ex/Regions.py [hunk 1]
```diff
@@ -167,16 +167,10 @@ def create_regions(world: MultiWorld, player: int):
     world.regions.append(regBitS)
 
 
-def connect_regions(world: MultiWorld, player: int, source: str, target: str, rule=None):
+def connect_regions(world: MultiWorld, player: int, source: str, target: str, rule=None, name: str = None):
```
Comment (N00byKing) [inline worlds/sm64ex/Regions.py:170]: iirc `connect_regions` is now available in one of the base classes right? If you're already changing it, could you just delete the function definition altogether and replace it with the new one?

Comment (alwaysintreble) [reply worlds/sm64ex/Regions.py:170]: They're using one of the new connect helpers, which is as close to the function you call there as we have. Moving them to use one of the new helpers directly would require a full rewrite of the file.

Comment (N00byKing) [reply worlds/sm64ex/Regions.py:170]: Nvm, I misread the change. What is done now looks exactly like I wanted.
Thats what I get for trying to read the diff from my phone ^^'

**Review by N00byKing — APPROVED** [review]

**Review by TRPG0 — APPROVED** [review]
Hylics 2 is fixed in #2460

**Comment by black-sliver** [comment]
I'm wondering how much overhead string building for each entrance for each player adds. Not the action of string concat itself, but the fact that we keep duplicates in memory and the strings are hot memory when used in region cache.

**Comment by Jarno458** [comment]
I do not like how this makes the spoiler log so much harder to read by the repetition of text
ClutteredSpoilerlog

**Comment by Jarno458** [comment]
My feedback was discussed with @alwaysintreble yesterday by dm's i cant say we reached an agreement, but 
using spaces for entrance names its an ugly work around, and using the default entrance names makes the paths harder to read

Two whole different approaches where discussed,
1. to remove topology and add custom paths, there is supposed to be PR for that somewhere, i am not a fan of that as i rather just have the core handle it for me, looked into the code used to create the paths and i dont feel confortable to duplicate the logic to a world specific implementation.

2. a second topology_present option like topology_present = 2, where it will print out the paths in the spoiler but without the entrance names. i also looked into the feasability of that, but the whole pathing code is confusing to me so it doesnt not seem easy to implement

**Comment by Jarno458** [comment]
There seems to be no simpel fix for this, i guess ill accept the degradation in spoiler log :/

**Review by Jarno458 — APPROVED** [review]

**Review by digiholic — APPROVED** [review]

**Review by kindasneaki — APPROVED** [review]
I looked and tested the new changes and they look good to me

**Review by black-sliver — APPROVED** [review]
Since the cache breaks on non-unique names, I think this is a blocker. I.e. if people load APWorlds they may silently break.

I'd like for @Berserker66 to double-check and hit merge though

**Comment by alwaysintreble** [comment]
had to make a change due to move rando getting merged and failing the new assert

**Review by N00byKing — APPROVED** [review]

**Review by Berserker66 — APPROVED** [review]

**Pull request merged**